### PR TITLE
[#1732] Enhance Redis data consolidation command with comprehensive migration tools

### LINKED ISSUE #1732: Revisit consolidate-redis-models command
Related to #1728 #1726 

Analyze the current implementation of `lib/onetime/cli/migrate_redis_data_command.rb` and enhance it with a more efficient approach. Particularly focusing on using MIGRATE for same-instance migrations and providing a robust fallback for cross-server scenarios.

### Meta-Consideration: Familia vs. Local Implementation

Consider the possibility of implementing this in Familia. Given that:
- Familia already has `Familia::Tools.move_keys`
- This migration logic is generic and reusable
- Other applications using Familia could benefit from optimized migration

**Recommendation:** The enhanced migration functionality SHOULD be implemented in Familia as `Familia::Tools::Migration` or enhanced `Familia::Tools.move_keys` with strategy detection. For now, we'll implement locally in OneTimeSecret as a proof of concept that can later be contributed upstream to Familia.

### Implementation Adjustments

1. **OT Module Integration**
   - Add `require_relative 'alias'` to properly load OT
   - Continue using `OT.li`, `OT.le`, `OT.ld` for consistent logging

2. **MIGRATE Command Approach**
   - Implement the MIGRATE logic BUT also generate redis-cli commands
   - Output commands for manual execution when verbose mode is enabled
   - Provide fallback instructions if MIGRATE fails

3. **Tryouts Test Structure**
   ```ruby
   # Setup section for shared instance variables
   @redis_host = 'localhost'
   @redis_port = 6379
   @test_db_source = 14
   @test_db_target = 15

   def setup_test_data
     # Helper methods accessible across test cases
   end

   ## Test case 1
   # Can now use @redis_host, etc.
   ```

4. **User-Executable Commands Output**
   - Generate and display redis-cli commands for:
     - Bulk MIGRATE operations
     - Manual DUMP/RESTORE sequences
     - Cleanup operations (user can run DELETE if desired)
   - Example output:
   ```bash
   # For same-instance migration (Redis 3.0.6+):
   redis-cli -n 14 --raw MIGRATE localhost 6379 "" 0 5000 COPY REPLACE KEYS customer:1 customer:2 customer:3

   # For cleanup (run at your own risk):
   redis-cli -n 14 DEL customer:1 customer:2 customer:3
   ```

### Updated File Structure

1. **`lib/onetime/redis_key_migrator.rb`**
   - Require `onetime/alias` for OT
   - Add method `generate_cli_commands` to output redis-cli commands
   - Keep MIGRATE implementation but add clear error messages with fallback commands

2. **`lib/onetime/cli/migrate_redis_data_command.rb`**
   - When verbose, output generated redis-cli commands
   - Add option `--show-commands` to only display commands without executing

3. **`try/redis_key_migrator_try.rb`**
   - Restructure with proper setup section
   - Instance variables in setup for test configuration
   - Helper methods defined in setup section

### Enhanced Migration Output Example

```
Migration Preview:
  • 1000 customer keys: DB 6 → DB 0
  Strategy: MIGRATE (same Redis instance detected)

Redis CLI Commands (for manual execution if needed):

  # Migration command:
  redis-cli -n 6 --raw MIGRATE localhost 6379 "" 0 5000 COPY REPLACE KEYS \
    customer:* [first 100 keys shown, use SCAN for full list]

  # To list all matching keys:
  redis-cli -n 6 --scan --pattern "customer:*"

  # To verify migration:
  redis-cli -n 0 DBSIZE
  redis-cli -n 0 --scan --pattern "customer:*" | head -10

  # Optional cleanup (removes from source):
  redis-cli -n 6 --scan --pattern "customer:*" | xargs redis-cli -n 6 DEL

Proceeding with automated migration...
[progress bar or output]
```

### Key Benefits of This Approach

1. **Transparency:** Users see exactly what commands are being run
2. **Fallback:** If automated migration fails, users have manual commands ready
3. **Education:** Users learn the underlying Redis commands
4. **Safety:** No automatic deletion; users must explicitly run cleanup
5. **Debugging:** Verbose mode shows exact commands for troubleshooting

### Future Familia Contribution

Once proven in OneTimeSecret, this implementation could be:

1. Extracted into a `Familia::Migration` module
2. PR'd to the Familia gem with:
   - Strategy detection (MIGRATE vs DUMP/RESTORE)
   - SCAN-based discovery
   - Batch processing
   - CLI command generation
   - Progress callbacks
3. OneTimeSecret then updated to use the new Familia version

This approach provides immediate value while setting up for a proper upstream contribution.

---

## Dev Notes

<details><summary>Possible Approaches</summary>
<p>

  ## Approach 1: MIGRATE for Same Instance, Different DB

  Starting with Redis 3.0.6 MIGRATE supports a new bulk-migration mode that uses pipelining in order to migrate multiple keys between instances without incurring in the round trip time latency and other overheads that there are when moving each key with a single 
  MIGRATE call.

  In order to enable this form, the KEYS option is used, and the normal key argument is set to an empty string. The actual key names will be provided after the KEYS argument itself, like in the following example:

  MIGRATE 1.2.3.4 6379 "" 0 5000 KEYS key1 key2 key3

  ## Approach 2: MIGRATE with SCAN

  ```ruby
  require 'redis'

  class RedisMigrator
    def initialize(source_host, source_port, source_db, target_host, target_port, target_db)
      @source = Redis.new(host: source_host, port: source_port, db: source_db)
      @target_host = target_host
      @target_port = target_port
      @target_db = target_db
    end

    def migrate_pattern(pattern = '*', batch_size = 100, timeout = 5000)
      cursor = 0
      total_migrated = 0

      loop do
        cursor, keys = @source.scan(cursor, match: pattern, count: 1000)

        unless keys.empty?
          keys.each_slice(batch_size) do |batch|
            begin
              @source.migrate(@target_host, @target_port, '', @target_db, timeout, 
                             keys: batch)
              total_migrated += batch.size
              puts "Migrated #{total_migrated} keys"
            rescue => e
              puts "Error with batch: #{e.message}"
              migrate_individual_keys(batch, timeout)
            end
          end
        end

        break if cursor == 0
      end

      puts "Migration complete. Total migrated: #{total_migrated}"
    end

    private

    def migrate_individual_keys(keys, timeout)
      keys.each do |key|
        begin
          @source.migrate(@target_host, @target_port, key, @target_db, timeout)
        rescue => e
          puts "Failed to migrate #{key}: #{e.message}"
        end
      end
    end
  end

  # Usage
  migrator = RedisMigrator.new('localhost', 6379, 0, 'localhost', 6379, 1)
  migrator.migrate_pattern('cache:*')
  ```

  ## Approach 3: Scan-Based Migration (Memory Efficient)
  For large datasets, use SCAN to avoid blocking:

  ```ruby
  require 'redis'

  class RedisMigrator
    def initialize(source_db, target_db, pattern = '*')
      @source = Redis.new(host: 'localhost', port: 6379, db: source_db)
      @target = Redis.new(host: 'localhost', port: 6379, db: target_db)
      @pattern = pattern
    end

    def migrate_all
      cursor = 0
      migrated_count = 0

      loop do
        cursor, keys = @source.scan(cursor, match: @pattern, count: 1000)

        keys.each do |key|
          migrate_single_key(key)
          migrated_count += 1
        end

        puts "Migrated #{migrated_count} keys so far..." if migrated_count % 1000 == 0

        break if cursor == 0
      end

      puts "Migration complete. Total keys migrated: #{migrated_count}"
    end

    private

    def migrate_single_key(key)
      dumped = @source.dump(key)
      return unless dumped

      ttl = @source.pttl(key)
      ttl = 0 if ttl == -1

      @target.restore(key, ttl, dumped, replace: true)
      @source.del(key)
    rescue => e
      puts "Error migrating #{key}: #{e.message}"
    end
  end

  # Usage
  migrator = RedisMigrator.new(6, 0, 'customer:*')
  migrator.migrate_all
  ```

  ## Approach 4: Pipeline for Batch Operations
  For better performance with many keys:

  ```ruby
  require 'redis'

  source_redis = Redis.new(host: 'localhost', port: 6379, db: 0)
  target_redis = Redis.new(host: 'localhost', port: 6379, db: 1)

  def batch_migrate(source, target, keys, batch_size = 100)
    keys.each_slice(batch_size) do |batch|
      # Pipeline DUMP operations
      dumps = source.pipelined do |pipe|
        batch.each { |key| pipe.dump(key) }
      end

      ttls = source.pipelined do |pipe|
        batch.each { |key| pipe.pttl(key) }
      end

      # Pipeline RESTORE operations
      target.pipelined do |pipe|
        batch.each_with_index do |key, idx|
          next if dumps[idx].nil?

          ttl = ttls[idx] == -1 ? 0 : ttls[idx]
          pipe.restore(key, ttl, dumps[idx], replace: true)
        end
      end

      # Pipeline DELETE operations
      source.pipelined do |pipe|
        batch.each { |key| pipe.del(key) }
      end

      puts "Migrated batch of #{batch.size} keys"
    end
  end

  keys = source_redis.keys('session:*')
  batch_migrate(source_redis, target_redis, keys)

  ```

</p>
</details>

### PR DESCRIPTION
**Pull request by delano**
Resolves #1732

## Summary

This PR significantly enhances the Redis data migration tooling in Onetime Secret, replacing the simple `migrate-redis-data` command with a comprehensive `consolidate-redis-models` system that provides safer, more flexible data consolidation capabilities.

The consolidated migration system includes intelligent strategy detection, improved error handling, and multiple execution modes for both automated and manual operations. It properly handles edge cases like Redis URI normalization and provides better UX through batch processing and detailed progress reporting.

### Command Interface Improvements
- **Renamed command**: `migrate-redis-data` → `consolidate-redis-models` for clarity
- **Enhanced CLI options**: Added `--show-commands`, `--batch-size`, `--yes` flags
- **Better help system**: Comprehensive usage documentation with examples
- **Safer defaults**: Requires explicit `--run` flag to prevent accidental execution

### Migration Strategy Intelligence  
- **Automatic strategy detection**: DUMP/RESTORE, MIGRATE, or COPY based on Redis version and configuration
- **Redis URI normalization**: Properly handles default port 6379 in connection strings
- **Connection optimization**: Reuses connections and validates before migration
- **CLI command generation**: Produces manual redis-cli commands for review/execution

### Enhanced Safety & Monitoring
- **Comprehensive validation**: Batch size limits (1-10,000), connection testing
- **Detailed progress tracking**: Real-time migration statistics and error reporting
- **Copy-first strategy**: Preserves source data during migration by default
- **Migration flags cleanup**: Automatic cleanup of detection flags post-migration

### Testing & Documentation
- **Extensive test coverage**: 6 dedicated tryout files covering unit, integration, and edge cases
- **Redis debug documentation**: New guide for enabling DEBUG POPULATE in development
- **Improved detection logic**: Enhanced legacy data scanning with migration mode support

### Code Organization
- **Modular architecture**: Separated `RedisKeyMigrator` class for reusability  
- **Command consolidation**: Removed confusing `move-keys` command to avoid overlap
- **Better error handling**: Comprehensive exception handling with detailed logging

## Migration Impact

Existing `bin/ots migrate-redis-data` installations will need to update to `bin/ots consolidate-redis-models`. The new command maintains backward compatibility for core functionality while providing significant improvements in safety and usability.

## Test Plan

- [x] All existing tryout tests pass
- [x] New migration strategy tests validate Redis version detection
- [x] Edge case handling for connection failures and invalid batch sizes
- [x] CLI command generation produces valid redis-cli syntax
- [x] Migration mode properly detects all non-DB-0 data
- [x] Batch processing handles large datasets efficiently
- [x] Progress reporting works in both verbose and quiet modes

🤖 Generated with [Claude Code](https://claude.ai/code)

___

### **PR Type**
Enhancement

___

### **Description**
- Replace `migrate-redis-data` with comprehensive `consolidate-redis-models` command

- Add intelligent Redis migration strategy detection and CLI generation

- Implement new `RedisKeyMigrator` class with batch processing capabilities

- Enhance legacy data detection with migration mode support

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Legacy migrate-redis-data"] --> B["New consolidate-redis-models"]
  B --> C["RedisKeyMigrator Class"]
  C --> D["Strategy Detection"]
  D --> E["MIGRATE Command"]
  D --> F["DUMP/RESTORE"]
  B --> G["CLI Generation"]
  G --> H["Manual Commands"]
  C --> I["Batch Processing"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>cli.rb</strong><dd><code>Update CLI command imports</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1735/files#diff-682e856ca4b6e4f9b379fa54b0f348a75da65b5f589f9da4b23de0f49cd6d1bb">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>ots</strong><dd><code>Update CLI command definitions and options</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1735/files#diff-f760b4e89ed099696292fc69306124cbec19b6c57ac9be1f5ff22bc4b887171b">+9/-7</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td><strong>consolidate_redis_models_command.rb</strong><dd><code>New comprehensive Redis consolidation command</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1735/files#diff-d77f2abe969c965cbb2db6203fa00d2e3a20906f940461da1f24085b97caa7e2">+320/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>detect_legacy_data.rb</strong><dd><code>Add migration mode for comprehensive detection</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1735/files#diff-39d5885d87a4c7925aa18d6ba98a029e7fe7517c2762e86cd89dcbbc46b9c4a9">+84/-49</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>redis_key_migrator.rb</strong><dd><code>New Redis key migration class implementation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1735/files#diff-dbf0d85cc76068e5146f8486965609380701f7bc90cca66998719624032ba918">+597/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Miscellaneous</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>migrate_redis_data_command.rb</strong><dd><code>Remove old migration command</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1735/files#diff-15fe85e72c67dda76cb8a86f7a87a24a8166971098d01c8838b6c859951ce5e3">+0/-163</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>move_keys_command.rb</strong><dd><code>Remove redundant move keys command</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1735/files#diff-b0078823a94078f67d68019cd318a25c105b430e8d9cace2ec3f72736bc49ec4">+0/-24</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>6 files</summary><table>
<tr>
  <td><strong>after_load_spec.rb</strong><dd><code>Remove environment assertion from test</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1735/files#diff-a1c69b1bb9fbafb8dfc824791857a274b0b0f0530b8bc7c1e6cf6ee90d8a5c24">+0/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>detect_legacy_data_try.rb</strong><dd><code>Update test for customdomain model name</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1735/files#diff-40993b7871d58e8577cdf134e036efecb840ec141779586800aa9f7aa4a65a11">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>redis_debug_try.rb</strong><dd><code>Add Redis debug testing utilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1735/files#diff-6a8682200318f49c86053323e8f47e10dc650ad557d845722d430b23209a3405">+30/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>redis_key_migrator_basic_try.rb</strong><dd><code>Add basic migrator functionality tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1735/files#diff-4319ed9d12bca4a9339f6ea21e68f6f5f21f783e17509f4e540fef9e166bc19b">+33/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>redis_key_migrator_integration_try.rb</strong><dd><code>Add comprehensive integration tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1735/files#diff-5c0c98bf5c0f7242b136e76103e296f7c08f637feef332b54ff5b4695d01acbf">+197/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>redis_key_migrator_unit_try.rb</strong><dd><code>Add unit tests with mock implementation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1735/files#diff-5fce3e73f9d7911173f8d2393fbf3041deb3627c2d92efba96235f4ba94bd478">+204/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>OVERVIEW.md</strong><dd><code>Minor documentation improvements</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1735/files#diff-963be3bebe8a33ecf48cfb1a2d90dd4d7f3c5c13577ee3f65aabb1c55c76d70d">+5/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>redis-debug.md</strong><dd><code>Add Redis DEBUG command configuration guide</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1735/files#diff-1f0650c423554d6eea26d11429368cd0d5439a1c0ec53a9e02f2f699978c162c">+56/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: lib/onetime/redis_key_migrator.rb [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+
+    def uris_identical?
+      source_normalized = normalize_uri(@source_uri)
+      target_normalized = normalize_uri(@target_uri)
+      source_normalized == target_normalized
+    end
+
+    def normalize_uri(uri)
+      port = uri.port || 6379
+      "#{uri.host}:#{port}/#{uri.path&.gsub('/', '') || uri.db || 0}"
```
[the remark this answers is not in the document]
Comment (delano) [reply lib/onetime/redis_key_migrator.rb:283]: URI::Redis objects had a db method
